src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/ProjectKey.cs

```diff
@@ -0,0 +1,52 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using Microsoft.AspNetCore.Razor;
+
+namespace Microsoft.CodeAnalysis.Razor.ProjectSystem;
+
+/// <summary>
+/// A very light wrapper around a file path, used to ensure consistency across the code base for what constitutes the unique
+/// identifier for a project.
+/// </summary>
+internal sealed class ProjectKey
+{
+    // ProjectKey represents the full path to the project file name. All creation logic is here in one place
+    // to ensure this is consistent.
+    public static ProjectKey From(HostProject hostProject) => new(hostProject.FilePath);
+    public static ProjectKey From(IProjectSnapshot project) => new(project.FilePath);
+    public static ProjectKey? From(Project project) => project.FilePath is null ? null : new(project.FilePath);
+    public static ProjectKey From(string projectFilePath)
+    {
+        // Right now, we expect this to be a project file path. This will change in future
+        Debug.Assert(projectFilePath.EndsWith(".csproj", StringComparison.OrdinalIgnoreCase));
+        return new(projectFilePath);
+    }
+    public static ProjectKey FromLegacy(string projectFilePath)
+    {
+        // In the legacy editor, we expect this to be a project file path.
+        Debug.Assert(projectFilePath.EndsWith(".csproj", StringComparison.OrdinalIgnoreCase));
+        return new(projectFilePath);
+    }
+
+    public string Id { get; }
+
+    private ProjectKey(string id)
+    {
+        Debug.Assert(id is not null, "Cannot create a key for null Id. Did you call ProjectKey.From(this) in a constructor, before initializing a property?");
```

Comment:
Interesting. Why not just do an `ArgumentNullException`?

Reply:
I wanted this to stand out. Because of nullable reference types, this can only occur when someone calls `ProjectKey.From(this)` in a constructor, before the properties have all been initialized, so this is just here to prevent future developer accidents.